feat(stats): Add heartbeat

erezrokah opened the pull request:
~~Draft PR on how we could implement heartbeat with different stats~~

Fixes https://github.com/cloudquery/cloudquery/issues/708

Uses [segment stats library](https://github.com/segmentio/stats) to measure durations. Only added a single measurement in `callTableResolve`, but it's easy to add more. Future improvements can be logging to console slow resources, sending information to [datadog](https://github.com/segmentio/stats#engine), etc

Example log for still running operation:
```
{"@level":"debug","@message":"heartbeat","@module":"aws","@timestamp":"2022-05-15T15:49:19.377174+03:00","id":"callTableResolve:client_id:6157xxxxxxxx:eu-north-1:table:aws_iam_roles","running_for":"2m2s"}
```

After the operation is done:
```
{"@level":"debug","@message":"heartbeat","@module":"aws","@timestamp":"2022-05-15T15:49:49.372747+03:00","duration":"2m9s","id":"callTableResolve:client_id:6157xxxxxxxx:eu-north-1:table:aws_iam_roles"}
```

erezrokah committed:
feat(stats): Add heartbeat

erezrokah commented on provider/execution/execution.go:
```diff
@@ -225,6 +226,9 @@ func (e TableExecutor) cleanupStaleData(ctx context.Context, client schema.Clien
 
 // callTableResolve does the actual resolving of the table calling the root table's resolver and for each returned resource resolves its columns and relations.
 func (e TableExecutor) callTableResolve(ctx context.Context, client schema.ClientMeta, parent *schema.Resource) (uint64, diag.Diagnostics) {
+	clock := stats.DefaultEngine.Clock("callTableResolve", stats.Tag{Name: "table", Value: e.Table.Name})
```
We can add more tags if we'd like.
`clock.Stop` add another measurement with the duration, so we can track what operations completed and how long they took

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,45 @@
+
+func (h *LogHandler) Flush() {
+	for m := range h.measures {
+		h.logger.Debug("heartbeat", "time", m.time, "measures", m)
```
We probably want something better here, like aggregation of the data

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+func Start(logger hclog.Logger) {
+	stats.Register(newHandler(logger))
```
We can register more handlers, like Datadog - see https://github.com/segmentio/stats#engine

erezrokah committed:
chore: Initial commit

erezrokah committed:
feat: Improve reporting

erezrokah committed:
refactor: Code cleanup

erezrokah committed:
test: Add tests

erezrokah committed:
fix: Change ticker interval to minute

roneli commented on stats/stats.go:
```diff
@@ -0,0 +1,140 @@
+
+	go func() {
+		ticker := time.NewTicker(opts.Tick)
+		for range ticker.C {
```
Moreover I think this will skew with line 94, as we do since after the measure it will be +1min from actual duration.

erezrokah replied:
By buffered do you mean initialize the channel with a size? If so see https://github.com/cloudquery/cq-provider-sdk/pull/237/files#diff-58dc02fad889eab26b4f7d1a7343ab0c7b8d2ffbb959a6c728fc77aa87b539afR109

If not can you explain a bit more?

>Moreover I think this will skew with line 94, as we do since after the measure it will be +1min from actual duration.

Can you explain this a bit? The duration is calculated once the `clock.Stop` method is called (the clock has an internal field that keeps track of the time)

roneli commented on stats/stats.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	durationReported := make([]string, 0)
+	for el := h.stats.Front(); el != nil; el = el.Next() {
```
why not use a queue structure rather then a map?

erezrokah replied:
We group 2 different measurements, `start` and `stop` by the same id, hence the usage of a map to access existing measurements.

A queue might not provide easy access to existing measurements.

The heartbeat keeps printing a running operation until a `clock.Stop` was called and then it will print the duration once.

roneli replied:
I am very confused fromcode as it only does a simple print every X this solution seems over complex. ill need to look again on the code and run it so I can give you better feedback.

erezrokah replied:
I'm happy to sync on this. It keeps track of the operations that were started and completed.
On every tick, it logs the tracked operations either with `running_for` if the operation wasn't completed or with `duration` if the operation was completed.
If the operation was completed is won't be printed again.

roneli reviewed (changes requested):
Great start, I like the usage of segments stat collector.

But I am not sure after re-reading the whole code again, you are doing some logic on the stats attribute, its not clear, nor documented on it's purpose, so my comments might be off on exactly stats/measures mean.

I think a queue structure would be smarter or just printing the measure once its received to the logger.

Moreover, this can be done in a separate PR, but we don't update in execution at the current position i.e where its currently at i.e (resolving resource/columns, inserting) etc'

disq commented on provider/execution/execution.go:
```diff
@@ -12,6 +12,8 @@ import (
 	"github.com/cloudquery/cq-provider-sdk/provider/diag"
 	"github.com/cloudquery/cq-provider-sdk/provider/schema"
+	"github.com/cloudquery/cq-provider-sdk/stats"
+	s "github.com/segmentio/stats/v4"
```
`segmentstats` or `sstats` maybe.

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,140 @@
+func (h *LogHandler) HandleMeasures(time time.Time, measures ...stats.Measure) {
+	for _, m := range measures {
+		id, stamp := meta(m.Fields[0].Name, m.Tags)
+		h.measures <- Measure{id: id, duration: m.Fields[0].Value.Duration(), time: time, stamp: stamp}
```
`HandleMeasures` runs in the context of the calling method, and `Flush` runs in the context of the go routine.
If an operation is stuck the `clock.Stop` will never be called, so we:
1. Won't get the duration measurement
2. Won't get any additional logs or heartbeat

erezrokah committed:
style: Fix style issues

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,140 @@
+	mu       sync.Mutex
+}
+
+type Options struct {
+	Logger  hclog.Logger
+	Tick    time.Duration
+	Handler stats.Handler
+}
```
Ah, thanks I followed the existing pattern in the code from https://github.com/cloudquery/cq-provider-sdk/blob/be60d7430a62f4b1d328c05b193ce55dd01c6fd1/serve/debug.go#L35.

Let me use the one you sent

erezrokah replied:
Update:

@roneli can you clarify how the function options pattern fit in when there is not constructor?
I can add one, but it might be confusing since this package is meant to be used without it.

erezrokah replied:
Ah, I think I understand, thanks for clarifying.
Is this https://github.com/cloudquery/cq-provider-sdk/pull/237/commits/67b759131cecb602d6a2226f3b59b776262f71d9 what you had in mind?

I kept the context as logger as arguments as those are required

erezrokah committed:
style: Fix style issues

erezrokah committed:
refactor: Add code comments

erezrokah commented:
@roneli Thanks for the detailed review. I've added some code comments in hope to make the logic easier to understand.

Also addressed the style issues (please let me know if something is missing), and added a question on the functional options pattern you suggested

erezrokah committed:
fix: Use go functional options pattern

erezrokah committed:
refactor: Rename stats to trackedOperations

erezrokah committed:
style: Use private members

erezrokah committed:
refactor: Use mutex instead of channel

erezrokah commented:
After syncing with @roneli we decided to drop the channel in favor a lock. This is done in https://github.com/cloudquery/cq-provider-sdk/pull/237/commits/6e8077f85e4fab714e1d702d2a968aa38f60cfc9

roneli commented on stats/stats.go:
```diff
@@ -0,0 +1,141 @@
+	return cl
+}
+
+func meta(name string, tags []stats.Tag) (string, bool) {
```
please sort the private and public methods in the file.

erezrokah replied:
Done in https://github.com/cloudquery/cq-provider-sdk/pull/237/commits/fc1fb250a852284e19df964d65da57165d5de25a

roneli reviewed (commented):
I suggest someone else review this as well. maybe @disq

erezrokah committed:
refactor: Move private method to the end

disq commented on stats/stats.go:
```diff
@@ -0,0 +1,141 @@
+
+func meta(name string, tags []stats.Tag) (string, bool) {
+	var stamp = false
+	var s []string
```
```suggestion
	s := make([]string, 0, len(tags))
```
or even
```suggestion
	s := make([]string, 1, len(tags)))
	s[0] = name
```

disq reviewed (changes requested):
I'll re-review once it's more documented :)

erezrokah committed:
refactor: Update stats/stats.go

Co-authored-by: Kemal <223029+disq@users.noreply.github.com>

erezrokah committed:
refactor: Update stats/stats.go

Co-authored-by: Kemal <223029+disq@users.noreply.github.com>

erezrokah committed:
refactor: Add size

erezrokah committed:
refactor: Replace range with case

erezrokah committed:
refactor: Rename to durationLogger

erezrokah committed:
refactor: Rename to getMeasurementId

erezrokah committed:
refactor: Add Stopped flag

erezrokah committed:
refactor: Add comments, change to private fields

erezrokah commented on provider/execution/execution.go:
```diff
@@ -225,6 +227,9 @@ func (e TableExecutor) cleanupStaleData(ctx context.Context, client schema.Clien
 
 // callTableResolve does the actual resolving of the table calling the root table's resolver and for each returned resource resolves its columns and relations.
 func (e TableExecutor) callTableResolve(ctx context.Context, client schema.ClientMeta, parent *schema.Resource) (uint64, diag.Diagnostics) {
+	clock := stats.NewClockWithObserve("callTableResolve", segmentStats.Tag{Name: "client_id", Value: identifyClient(client)}, segmentStats.Tag{Name: "table", Value: e.Table.Name})
```
It can take and empty string

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,141 @@
+		id := el.Key
+		stat := el.Value.(stat)
+		if stat.Duration == 0 {
+			// `clock.Stop` was not called, so the operation is still running
```
Added a `stopped` flag

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,141 @@
+
+	go func() {
+		ticker := time.NewTicker(opts.tick)
+		for range ticker.C {
```
Done in https://github.com/cloudquery/cq-provider-sdk/pull/237/commits/566b16c45810edbc0292839aa5872b69bd19c25b

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,141 @@
+	Duration time.Duration
+}
+
+type logHandler struct {
```
Done in https://github.com/cloudquery/cq-provider-sdk/pull/237/commits/bd1feb6133c2f368b420871879f3af6ee7c2c461

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,141 @@
+	return &logHandler{logger: logger, trackedOperations: orderedmap.NewOrderedMap()}
+}
+
+func meta(name string, tags []stats.Tag) (string, bool) {
```
Renamed in https://github.com/cloudquery/cq-provider-sdk/pull/237/commits/4949ef838dcbdfe2f49905e6e8593db4d8287f3a

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,146 @@
+				h.trackedOperations.Set(id, stat{start: item.(stat).start, duration: m.Fields[0].Value.Duration(), stopped: true})
+			}
+		} else {
+			h.trackedOperations.Set(id, stat{start: time, duration: 0, stopped: false})
```
Not sure what is the style we want. We could use https://github.com/GaijinEntertainment/go-exhaustruct to enforce initializing structs

disq replied:
no need, zero value is valid in Go.
```suggestion
			h.trackedOperations.Set(id, stat{start: time})
```

erezrokah commented:
Looking at the failed tests

erezrokah committed:
test: Fix tests

disq reviewed (approved):
Looks better, still some nits

erezrokah committed:
refactor: Remove struct init

erezrokah commented on stats/stats.go:
```diff
@@ -0,0 +1,146 @@
+		if stat.stopped {
+			// `clock.Stop` was called, so we log the total duration and remove the operation from future logs
+			durationReported = append(durationReported, id.(string))
+			h.logger.Debug("heartbeat", "id", id, "duration", stat.duration.Round(time.Second).String())
```
Yes, I wanted to drop the `ms` as it is not very useful.
Also it creates a non uniform format:
![image](https://user-images.githubusercontent.com/26760571/168852365-96aa1316-c2de-4bfd-87b9-0801587f3d6c.png)

disq replied:
it's always not-uniform if you keep doing `.String()`. To get seconds, you need `.Seconds()` (float64)
```suggestion
			h.logger.Debug("heartbeat", "id", id, "duration", int64(stat.duration.Round(time.Second).Seconds()))
```

erezrokah replied:
Maybe https://github.com/cloudquery/cq-provider-sdk/pull/237/commits/995ccf56f74e07d8c1d7497698a15e133b6ac02a is better?

erezrokah committed:
refactor: Always use seconds

Co-authored-by: Kemal <223029+disq@users.noreply.github.com>

erezrokah committed:
refactor: Better seconds logging

disq commented on stats/stats.go:
```diff
@@ -144,3 +145,7 @@ func getMeasurementDetails(name string, tags []stats.Tag) (string, bool) {
 	return strings.Join(s, ":"), stamp
 }
+
+func printSeconds(duration time.Duration) string {
```
`formatSeconds` ? it's not printing :)

erezrokah committed:
refactor: Rename to formatSeconds

erezrokah committed:
chore: Fix style issues

erezrokah committed:
refactor: Rename test method

roneli commented on serve/serve.go:
```diff
@@ -95,6 +98,7 @@ func Serve(opts *Options) {
 		})
 	}
 
+	stats.Start(context.Background(), opts.Logger)
```
nit: maybe add a log line that stats heartbeat collector started?

erezrokah replied:
Good point, done in https://github.com/cloudquery/cq-provider-sdk/pull/237/commits/3b394523d4b57c54e74f9e92eac122db777b7e76

roneli reviewed (approved):
LGTM, minor nit.

erezrokah committed:
fix: Add log message
